src/Polysemy/Internal.hs

```diff
@@ -534,6 +540,143 @@ subsumeUsing pr =
 {-# INLINE subsumeUsing #-}
 
 
+------------------------------------------------------------------------------
+-- | Finds the effect @e@ in the stack and inserts the effects @es@ after it.
+--
+-- @since TODO
+type family InsertedAfter (e :: Effect) (es :: EffectRow) (r :: EffectRow) :: EffectRow where
+  InsertedAfter e es (e : r) = e : Append es r
+  InsertedAfter e es (e0 : r) = e0 : InsertedAfter e es r
+
+------------------------------------------------------------------------------
+-- | Introduce a set of effects into 'Sem' after the position of the provided
+-- effect @e@. This is intended to be used with a type application:
+--
+-- @
+-- let
+--   sem1 :: Sem [e1, e2, e3, e4, e5] a
+--   sem1 = insertAfter @e2 @[e3, e4] (sem0 :: Sem [e1, e2, e5] a)
+-- @
+--
+-- @since TODO
+insertAfter ::
+  ∀ e es r r' a .
+  r' ~ InsertedAfter e es r =>
+  Subsume r r' =>
+  Sem r a ->
+  Sem r' a
+insertAfter = subsume_
+{-# inline insertAfter #-}
+
+------------------------------------------------------------------------------
+-- | Finds the effect @e@ in the stack and inserts the effects @es@ before it.
+--
+-- @since TODO
+type family InsertedBefore (e :: Effect) (es :: EffectRow) (r :: EffectRow) :: EffectRow where
+  InsertedBefore e es (e : r) = Append es (e : r)
+  InsertedBefore e es (e0 : r) = e0 : InsertedBefore e es r
+
+------------------------------------------------------------------------------
+-- | Introduce a set of effects into 'Sem' before the position of the provided
+-- effect @e@. This is intended to be used with a type application:
+--
+-- @
+-- let
+--   sem1 :: Sem [e1, e2, e3, e4, e5] a
+--   sem1 = insertBefore @e5 @[e3, e4] (sem0 :: Sem [e1, e2, e5] a)
+-- @
+--
+-- @since TODO
+insertBefore ::
+  ∀ e es r r' a .
+  r' ~ InsertedBefore e es r =>
+  Subsume r r' =>
+  Sem r a ->
+  Sem r' a
+insertBefore = subsume_
+{-# inline insertBefore #-}
+
+
+
+data SList l where
+  SEnd  :: SList '[]
+  SCons :: SList xs -> SList (x ': xs)
+
+class KnownList l where
+  singList :: SList l
+
+instance KnownList '[] where
+  singList = SEnd
+  {-# INLINE singList #-}
+
+instance KnownList xs => KnownList (x ': xs) where
+  singList = SCons singList
+  {-# INLINE singList #-}
```

Comment:
We should put all of the `SList` and `KnownList` boilerplate in a different module; I suggest `Polysemy.Internal.Membership`.
Also, replace the existing `KnownList` representation in `Polysemy.Internal.Bundle` with usages of this new one, instead.